bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -4379,6 +4490,627 @@ void ReplicaImp::setConflictDetectionBlockId(const ClientRequestMsg &clientReqMs
+
+    if (ps_) ps_->setCheckpointMsgInCheckWindow(lastExecutedSeqNum, checkMsg);
+
+    if (checkInfo.isCheckpointCertificateComplete()) {
+      onSeqNumIsStable(lastExecutedSeqNum);
+    }
+    checkInfo.setSelfExecutionTime(getMonotonicTime());
+    if (checkInfo.isCheckpointSuperStable()) {
+      onSeqNumIsSuperStable(lastStableSeqNum);
+    }
+
+    CryptoManager::instance().onCheckpoint(checkpointNum);
+  }
+}
+
+void ReplicaImp::updateLimitsAndMetrics(PrePrepareMsg *ppMsg) {
+  // TODO(GG): clean the following logic
+  if (mainLog->insideActiveWindow(lastExecutedSeqNum)) {  // update dynamicUpperLimitOfRounds
+    const SeqNumInfo &seqNumInfo = mainLog->get(lastExecutedSeqNum);
+    const Time firstInfo = seqNumInfo.getTimeOfFisrtRelevantInfoFromPrimary();
+    const Time currTime = getMonotonicTime();
+    if ((firstInfo < currTime)) {
+      const int64_t durationMilli = duration_cast<milliseconds>(currTime - firstInfo).count();
+      dynamicUpperLimitOfRounds->add(durationMilli);
+    }
+  }
+
+  if (config_.getdebugStatisticsEnabled()) {
+    DebugStatistics::onRequestCompleted(false);
+  }
+
+  auto numOfRequests = ppMsg->numberOfRequests();
+  if (numOfRequests > 0) {
+    SeqNumInfo &seqNumInfo = mainLog->get(lastExecutedSeqNum);
+    consensus_time_.add(seqNumInfo.getCommitDurationMs());
+    consensus_avg_time_.Get().Set((uint64_t)consensus_time_.avg());
+    if (consensus_time_.numOfElements() == 1000) consensus_time_.reset();  // We reset the average every 1000 samples
+    metric_last_executed_seq_num_.Get().Set(lastExecutedSeqNum);
+    metric_total_finished_consensuses_++;
+    if (seqNumInfo.slowPathStarted()) {
+      metric_total_slowPath_++;
+      if (numOfRequests > 0) {
+        metric_total_slowPath_requests_ += numOfRequests;
+      }
+    } else {
+      metric_total_fastPath_++;
+      if (numOfRequests > 0) {
+        metric_total_fastPath_requests_ += numOfRequests;
+      }
+    }
+  }
+}
+
+void ReplicaImp::handleDeferredRequests() {
+  if (isStartcollectingState_) {
+    if (!stateTransfer->isCollectingState()) {
+      LOG_INFO(GL, "Call to startCollectingState()");
+      time_in_state_transfer_.start();
+      clientsManager->clearAllPendingRequests();  // to avoid entering a new view on old request timeout
+      stateTransfer->startCollectingState();
+    } else {
+      LOG_ERROR(GL, "Collecting state should be active while we are in onFinishtExecuting");
+    }
+    isStartcollectingState_ = false;
+  }
+  if (!isCollectingState()) {
+    if (isSendCheckpointIfNeeded_) {
+      sendCheckpointIfNeeded();
+      isSendCheckpointIfNeeded_ = false;
+    }
+    if (isGoToNextView_) {
+      GotoNextView();
+      isGoToNextView_ = false;
+      isTryToGoNextView_ = false;
+    }
+    if (isTryToGoNextView_ && !isGoToNextView_) {
+      tryToGotoNextView();
+      isTryToGoNextView_ = false;
+    }
+    for (size_t i = 0; i < deferredMessages_.size(); i++) {
```

Comment:
You can probably use a modern syntax: for (const auto &msg : deferredMessages_)

Reply:
Done, thanks.